ledger/accountdb.go

```diff
@@ -900,31 +933,86 @@ func (qs *accountsDbQueries) lookupCreator(cidx basics.CreatableIndex, ctype bas
 // be retrieved.
 func (qs *accountsDbQueries) lookup(addr basics.Address) (data dbAccountData, err error) {
 	err = db.Retry(func() error {
-		var buf []byte
-		var rowid sql.NullInt64
-		err := qs.lookupStmt.QueryRow(addr[:]).Scan(&rowid, &data.round, &buf)
-		if err == nil {
-			data.addr = addr
-			if len(buf) > 0 && rowid.Valid {
-				data.rowid = rowid.Int64
-				return protocol.Decode(buf, &data.pad)
-			}
-			// we don't have that account, just return the database round.
-			return nil
+		data, err = lookupImpl(qs.lookupStmt, addr)
+		return err
+	})
+
+	return
+}
+
+func lookupImpl(lookupStmt *sql.Stmt, addr basics.Address) (data dbAccountData, err error) {
+	var buf []byte
+	var rowid sql.NullInt64
+	err = lookupStmt.QueryRow(addr[:]).Scan(&rowid, &data.round, &buf)
+	if err == nil {
+		data.addr = addr
+		if len(buf) > 0 && rowid.Valid {
+			data.rowid = rowid.Int64
+			err = protocol.Decode(buf, &data.pad)
+			return data, err
 		}
+		// we don't have that account, just return the database round.
+		return data, nil
+	}
 
-		// this should never happen; it indicates that we don't have a current round in the acctrounds table.
-		if err == sql.ErrNoRows {
-			// Return the zero value of data
-			return fmt.Errorf("unable to query account data for address %v : %w", addr, err)
+	// this should never happen; it indicates that we don't have a current round in the acctrounds table.
+	if err == sql.ErrNoRows {
+		// Return the zero value of data
+		return data, fmt.Errorf("unable to query account data for address %v : %w", addr, err)
+	}
+
+	return
+}
+
+func lookupExt(rdb db.Accessor, addr basics.Address, extension func(*sql.Stmt, *ledgercore.PersistedAccountData) error) (data dbAccountData, err error) {
+	rdb.Atomic(func(ctx context.Context, tx *sql.Tx) error {
```

Comment:
You'll need to add `err = rdb.Atomic` in order to catch `Commit` errors as well.

Reply:
Also, I think that you'd like to use `db.Retry` here.